### PR TITLE
[Scheduling] Set up infrastructure to model the basic problem.

### PR DESCRIPTION
**Pull request by jopperm**
This PR sets up the static scheduling infrastructure in `circt/Scheduling`. Only the `Scheduler` class, which models the basic HLS-style scheduling problem, is included here. The ASAP scheduler and a test pass will follow in separate PRs.

### CODE REVIEW

**Comment by jopperm** [comment]
@mikeurbach and/or @hanchenye, would you be willing to review this PR, whenever you've got some time?

**Comment by mikeurbach** [comment]
Yeah, sorry for the delay, but I will take a look.

**Comment by hanchenye** [comment]
Sorry, will take a look today!

File: include/circt/Scheduling/Scheduler.h [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+
+  Operation *getSource() { return src; }
+  unsigned getSourceIndex() { return srcIdx; }
+
+  Operation *getDestination() { return dst; }
+  unsigned getDestinationIndex() { return dstIdx; }
+
+private:
+  Dependence() = default;
+  Operation *src, *dst;
```
Comment (teqdruid) [inline include/circt/Scheduling/Scheduler.h:226]: See above comment.

File: include/circt/Scheduling/Scheduler.h [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+  //===--------------------------------------------------------------------===//
+public:
+  // Attempt to schedule the constructed problem.
+  virtual LogicalResult schedule() = 0;
+};
+
+/// This class models a *dependence* from the `srcIdx`'th result of `src` to the
+/// `destIdx`'th operand of `dst`. In other words, it represents an edge in a
+/// scheduling problem.
+class Dependence : public mlir::StorageUniquer::BaseStorage {
```
Comment (teqdruid) [inline include/circt/Scheduling/Scheduler.h:206]: I think this can be modeled by a tuple of `OpResult`, `OpOperand`.

File: include/circt/Scheduling/Scheduler.h [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+/// This class models the most basic scheduling problem.
+///
+/// A problem instance is comprised of:
+///
+///  - *Operations*: The vertices in the data-flow graph to be scheduled. We use
+///    the abbreviation `op` in the scheduling API.
+///  - *Dependences*: The edges in the data-flow graph to be scheduled, modeling
+///    a precedence relation between the involved operations. Individual operand
+///    and result indices can be distinguished. Abbreviation: `dep`
+///  - *Operator types*: An abstraction of the characteristics of the target
```
Comment (teqdruid) [inline include/circt/Scheduling/Scheduler.h:42]: What is the rationale for modeling operator types on top of the `Operation`s themselves? Why not just schedule the lowered operations themselves, lowered to the appropriate level first?

File: include/circt/Scheduling/Scheduler.h (hunk 2 above)
Comment (teqdruid) [reply include/circt/Scheduling/Scheduler.h:206]: Also, how can this be used to model out-of-SSA dependencies? (e.g. memory loads/stores) Or are you not worrying about that yet?

File: include/circt/Scheduling/Scheduler.h (hunk 3 above)
Comment (teqdruid) [reply include/circt/Scheduling/Scheduler.h:42]: If this is intended to describe the static `Op` instead of the dynamic `Operation`, look at `AbstractOperation`.

File: include/circt/Scheduling/Scheduler.h [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+///    an operator type. Abbreviation: `opr`
+///
+/// All components of the problem (ops, deps, oprs, as well as the problem
+/// itself) can be annotated with *properties*. In this basic problem, we model
+///
+/// - `assocOpr`, an operation-property that stores the operator type associated
+///   with each operation.
+/// - `latency`, an operator type-property denoting the number of time steps
+///   after which the operator's result is available.
+/// - `startTime`, an operation-property for the time step in which an operation
```
Comment (teqdruid) [inline include/circt/Scheduling/Scheduler.h:54]: Would it make sense for the output to be a [`PipelineOp`](https://github.com/llvm/circt/blob/012d69a27e4e11011c6c26ca9927a3a312af4b89/include/circt/Dialect/StaticLogic/StaticLogic.td#L32)?

**Review by teqdruid — COMMENTED** [review]
Sorry for the delay. I have some feedback on your modeling. It all makes sense, but I think you've re-invented a few things which are already part of MLIR.

File: include/circt/Scheduling/Scheduler.h (hunk 3 above)
Comment (mikeurbach) [reply include/circt/Scheduling/Scheduler.h:42]: If I am understanding the talk correctly, these operator types are the lower-level that will actually carry out the operations being scheduled, and can have properties of their own. If that's correct, I think I agree with @teqdruid that this kind of sounds like you could model them as `Operation`s. An `Operation` has a name and we use an `Operation *` to represent an opaque handle in lots of places.

I guess let me ask a question slightly differently: what kind of operator types could you forsee that _wouldn't_ make sense to model as operations in a dialect?

File: include/circt/Scheduling/Scheduler.h [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+  unsigned getStartTime(Operation *op) { return startTime.lookup(op); }
+
+  //===--------------------------------------------------------------------===//
+  // Hooks to check/verify the different problem components
+  //===--------------------------------------------------------------------===//
+protected:
+  virtual LogicalResult checkOp(Operation *op);
+  virtual LogicalResult checkDep(Dependence *dep);
+  virtual LogicalResult checkOpr(OperatorType *opr);
+  virtual LogicalResult checkProb();
```
Comment (mikeurbach) [inline include/circt/Scheduling/Scheduler.h:182]: Maybe spell out these abbreviations? `Op` is often abbreviated in MLIR projects, so I might not care about that, but the extra characters are worth it to me for the other ones. I know that `Prob` means `Problem` here, but I first read `Probability`, for example.

File: lib/Scheduling/Scheduler.cpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+
+OperatorType *Scheduler::getOrInsertOperatorType(StringRef name) {
+  auto *opr = uniquer.get<OperatorType>({}, name);
+  oprs.insert(opr);
+  return opr;
+}
+
+LogicalResult Scheduler::checkOp(Operation *op) {
+  if (!hasAssociatedOperatorType(op)) {
+    op->emitError("Operation is not associated with an operator type");
```
Comment (mikeurbach) [inline lib/Scheduling/Scheduler.cpp:40]: FYI you can `return op->emitError(...)` in failure cases. This will also make this a single line `if` so you could remove the curly braces. This comment applies to several similar cases below.

File: lib/Scheduling/Scheduler.cpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+    return failure();
+  }
+  return success();
+}
+
+LogicalResult Scheduler::checkProb() { return success(); }
+
+/// Check overall problem by delegating to the component-specific checkers.
+LogicalResult Scheduler::check() {
+  for (auto *op : getOperations()) {
```
Comment (mikeurbach) [inline lib/Scheduling/Scheduler.cpp:72]: nit: these for loops don't need the curly braces.

File: lib/Scheduling/Scheduler.cpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+  }
+
+  return verifyProb();
+}
+
+Dependence *
+Dependence::construct(mlir::StorageUniquer::StorageAllocator &allocator,
+                      const KeyTy &key) {
+  auto *result = allocator.allocate<Dependence>();
+  std::tie(result->src, result->dst, result->srcIdx, result->dstIdx) = key;
```
Comment (mikeurbach) [inline lib/Scheduling/Scheduler.cpp:146]: I don't know how I feel about `std::tie`. I usually avoid it, but this is a case where I actually think it might be worth it :thinking:

**Review by mikeurbach — COMMENTED** [review]
I think the code here looks pretty high-quality, barring some minor nits and MLIR-isms. For the most part, I agree with the previous comments and would like to hear a little more rationale/discussion. But I think after an iteration or two of discussion this will be ready to land!

File: include/circt/Scheduling/Scheduler.h [hunk 9]
```diff
@@ -0,0 +1,253 @@
...
+
+#include "circt/Support/LLVM.h"
+
+#include "mlir/Support/StorageUniquer.h"
+
+#include "llvm/ADT/DenseMap.h"
+#include "llvm/ADT/SetVector.h"
+
+namespace circt {
+namespace sched {
```
[the remark this answers is not in the document]
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h]: Done.

File: include/circt/Scheduling/Scheduler.h (hunk 5 above)
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h]: I got rid of the abbreviations everywhere, except in the parameter names -- is that ok?

File: lib/Scheduling/Scheduler.cpp (hunk 6 above)
Comment (jopperm) [reply lib/Scheduling/Scheduler.cpp]: Nice, didn't know that. Changed.

File: lib/Scheduling/Scheduler.cpp (hunk 7 above)
Comment (jopperm) [reply lib/Scheduling/Scheduler.cpp]: Done.

File: include/circt/Scheduling/Scheduler.h (hunk 4 above)
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h:53]: I think the interface should be dialect-agnostic at this level. Constructing a `PipelineOp` from the computed start times sounds like a great candidate for a utility function, if that becomes a common need for HLS flows in the future.

File: include/circt/Scheduling/Scheduler.h (hunk 3 above)
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h:41]: My rationale for modeling the the operator types explicitly is:
- they group together characteristics (latency, blocking time, propagation delay, ...) that otherwise would need to be attached to every operation.
- they are an elegant way to express that a subset of operations compete for the same resource. For example, let's say the target datapath will have two add-operators, and we want to execute operations *a*, *b* and *c* on them. In this scenario, we'd model one operator type "adder" with a limit of 2, and link it to *a*, *b*, *c*.
- they enable a scheduler to choose from a set of operator types (e.g. the small & slow vs. the large & fast multiplier).

Using the `AbstractOperation` to represent the operator type might be too inflexible. A `DivFOp` could be linked to different operator types based on the float type (f32 -> latency = 32, f64 -> latency = 64), or you could have a combined AddSub operator that would be used by `AddIOp`s and `SubIOp`s.

Using `Operation *`s as handles for the operator types would work great for "heavy things" like an FP operation, where you'd have a `hw.module[.extern|.generated]` that represents the operator type, but not so well for operations that just map to HDL primitives.

My `OperatorType` class gives clients the most freedom for modeling their operator libraries, but uses an extra class and the storage uniquer machinery just for representing a simple handle.

As I type this, I wonder if I should just use a StringRef? That would give us the good bits of the three alternatives above, at the cost of more expensive hashing when accessing the operator type property maps.

File: include/circt/Scheduling/Scheduler.h (hunk 2 above)
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h:209]: > Or are you not worrying about that yet?

No, quite the opposite, I'd really like to have a uniform modeling for the edges of the SSA graph as well as for memory dependencies. I played around a bit with `OpResult`/`OpOperand`, but these seem tightly knit into the SSA graph, and I currently don't see how that could work for memory dependences.

File: lib/Scheduling/Scheduler.cpp (hunk 8 above)
Comment (jopperm) [reply lib/Scheduling/Scheduler.cpp:137]: Saves 3 LoC here 😉 I'll revisit this after the discussion of the representation of dependences is resolved.

**Comment by jopperm** [comment]
Thanks for the reviews! The new commits address the cosmetic issues. I commented on the open design questions in the inline discussions.

File: include/circt/Scheduling/Scheduler.h (hunk 3 above)
Comment (mikeurbach) [reply include/circt/Scheduling/Scheduler.h:41]: I think I'm starting to understand why it makes sense to model these outside the Operation/AbstractOperation world. Giving clients freedom in this case seems like a good thing. Thanks for the explanation.

> As I type this, I wonder if I should just use a StringRef?

What about an `Identifier`? Those are interned within a context, and you could use the `containingOp`s context to get `Identifiers`, and then the lookups should be more efficient.

File: include/circt/Scheduling/Scheduler.h (hunk 4 above)
Comment (mikeurbach) [reply include/circt/Scheduling/Scheduler.h:53]: > I think the interface should be dialect-agnostic at this level.

I meant to comment on this, I agree with @jopperm here. It _could_ make sense to use a PipelineOp, but this Scheduling.h infrastructure probably doesn't want to force that.

File: include/circt/Scheduling/Scheduler.h (hunk 5 above)
Comment (mikeurbach) [reply include/circt/Scheduling/Scheduler.h]: Works for me, thanks! I was mainly concerned about the abbreviated method names.

File: include/circt/Scheduling/Scheduler.h (hunk 2 above)
Comment (teqdruid) [reply include/circt/Scheduling/Scheduler.h:209]: Yes, `OpResult`/`OpOperand` are married to the SSA graph, but so does this. By specifying the idx for the src and dst, you are replicating the functionality of them (literally -- look at their implementations). I'm also not wild about having duplicate def-use specifications -- what if they get out of sync?

I suggest a public API which allows clients to inform the scheduler about the non-SSA dependences of side-effectful operations. (Which could be represented by (src,dst) operation pairs.) It would then be an implementation detail (possibly with some private helper functions) which extract and add the SSA dependencies. I think that would be a more appropriate place for a unified dependence model.

File: include/circt/Scheduling/Scheduler.h (hunk 4 above)
Comment (teqdruid) [reply include/circt/Scheduling/Scheduler.h:53]: Agreed

File: include/circt/Scheduling/Scheduler.h (hunk 3 above)
Comment (teqdruid) [reply include/circt/Scheduling/Scheduler.h:41]: Let me rephrase your response to ensure that I'm understanding it: the reason for `OperatorType` is that there is a many-to-many mapping of operations to possible implementations of said operations wherein the scheduler may want to reason about resource contention?

If I'm interpreting you correctly, I _think_ it is reasonable to require clients model the possible implementations as operations themselves. E.g. an `AddSubOp` to model your resource which is capable of mapping both adds and subs into it. You could then use the `AddSubOp`s `AbstractOperation` instead of your `OperationType`. You could use it as a key for lookups (as you do with `OperationType` now), have default latencies (for well known operations e.g. in the comb dialect), and allow "implementation" operations to define their latencies as inherent properties.

I suspect this would be useful down the road as well -- for instance, a scheduler could output a mapping of `Operation*` to `AbstractOperation` showing which implementation it selected -- from there it would be trivial to lower it. Alternatively, the scheduler could actually instantiate the "implementation" `AbstractOperation`s into the `Operation*` and provide a mapping of `Operation*`s to schedule into the implementing `Operation*`s on which they should run and the schedule itself. A later 'multiplexing' pass could then wire up the correct multiplexing logic.

Does this sound reasonable?

File: include/circt/Scheduling/Scheduler.h (hunk 2 above)
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h:209]: Thanks for clarifying! I like that idea and will rework the API.

File: include/circt/Scheduling/Scheduler.h (hunk 3 above)
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h:41]: > Let me rephrase your response to ensure that I'm understanding it: the reason for `OperatorType` is that there is a many-to-many mapping of operations to possible implementations of said operations wherein the scheduler may want to reason about resource contention?

Correct, though I should note that in the beginning, we'll only support a many (Operations)-to-one (OperatorType) mapping. I do have an ILP-based scheduler to solve the many-to-many-problem, but it is only applicable if the number of ops to choose for is small. Developing a good and more scalable heuristic is what I'll look into next, research-wise.
 
> [...] Does this sound reasonable?

Makes a lot of sense.
- Operation *i* (to be scheduled) = `mlir::Operation`
- Operator type *q* = `mlir::AbstractOperation`
- Operator (an instance of *q*, executing *i* in a specific timestep) = `mlir::Operation`, i.e. instance of the `AbstractOperation`

This would require that clients model their operator library as `Operation`s, which might become quite tedious if they have modules with parameterized bit widths, for example. There was an ODM talk about [dynamic definition](https://llvm.discourse.group/t/open-mlir-meeting-6-17-2021-dynamic-definition-of-dialects-ops-types-attributes/3670) of dialects and ops (which I haven't watched yet), but maybe that'll become handy here.

The infra you outline in the last paragraph sounds amazingly useful.

Nevertheless, the scheduler class' interface currently is deliberately low-level with its imperative API (`addOperation`, `setLatency`, ...), and in this context, I think the `Identifier` representation of operator types would be good fit.

We can build the infrastructure you propose on top of that API (operator type's ID is the `AbstractOperation`'s name). Over time, we can remove the indirection via the `Identifier` (algorithms won't care, as its just a handle), and gradually design the necessary `[AbstractOperation|Op]Interface`s to query problem parameters directly from the target ops (instead of having someone `set(...)` them).

@teqdruid Would the `Identifier` approach acceptable to you to, as a short term solution to get the scheduling library off the ground?

Comment (teqdruid) [reply include/circt/Scheduling/Scheduler.h:41]: > This would require that clients model their operator library as `Operation`s, which might become quite tedious if they have modules with parameterized bit widths, for example.

Yeah, I could see that. Especially if they want to lower to a set of extern modules. What we really want is an AbstractOperation which can optionally store MLIR attributes to be added to the operation to be created _and_ could provide information about the implementation (e.g. latency).

> @teqdruid Would the Identifier approach acceptable to you to, as a short term solution to get the scheduling library off the ground?

Yeah, I'm OK with that and the flat API for now. We should plan on iterating it.

**Comment by jopperm** [comment]
I reworked the modeling of dependences, as @teqdruid suggested:
- Clients no longer need to "copy" the SSA graph into the problem. Instead, I automatically consider all def-use relationships between *registered* operations as dependences (distinguishing result and operand numbers).
- Clients can also add *auxiliary* dependences between ops, e.g. for memory dependences. (I think there will be more uses for this, so I didn't call them *memory* deps straight away.)
- Algorithms transparently iterate over both kinds of deps via a custom iterator.
- To be able to uniformly access properties on dependences, the API assigns numeric IDs to def-use and auxiliary deps on-demand. I expect most dependences to carry no properties, so I'd argue this is just a minor inconvenience. Maybe someone knows a better way to do this in C++?

File: include/circt/Scheduling/Scheduler.h (hunk 3 above)
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h:39]: > Yeah, I'm OK with that and the flat API for now. We should plan on iterating it.

Great. Yes, definitely, I'm curious to see how this evolves.

File: lib/Scheduling/Scheduler.cpp (hunk 8 above)
Comment (jopperm) [reply lib/Scheduling/Scheduler.cpp]: My need for the `tie` went away 🤓

File: include/circt/Scheduling/Scheduler.h (hunk 2 above)
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h]: Done (I think), see the new commits and my comment below (in the main conversation stream).

**Review by teqdruid — APPROVED** [review]
This looks great! Thanks for iterating.

I don't understand the use-case for `DependenceId`, but if it's just a key into a client's table, I would reduce the size of `Dependence` and use that instead. `getOrAssignDependenceId` would then just construct and check the `Dependence` before returning it. You'd have to add a few functions to make `Dependence` work as a key, so it's not as flexible as `unsigned` but I just hate redundancy in data structures.

One thing to keep in mind going forward is that this latency model is simple. Like @stephenneuendorffer was saying in the discourse forum, we'd really like to see a richer latency model. (i.e. to support device primitives with non-uniform input / output cycle requirements.) I've been participating in some internal design exercises to better develop an understanding about this and how we should model state, but I have no idea (yet!) what a more complex latency model should look like.

File: include/circt/Scheduling/Scheduler.h [hunk 10]
```diff
@@ -0,0 +1,254 @@
...
+  // Client access to properties: specify problem, retrieve solution
+  //===--------------------------------------------------------------------===//
+public:
+  void setLinkedOperatorType(Operation *op, OperatorType opr) {
+    linkedOperatorType[op] = opr;
+  }
+
+  void setLatency(OperatorType opr, unsigned val) { latency[opr] = val; }
+
+  Optional<unsigned> getStartTime(Operation *op) {
```
[the remark this answers is not in the document]
Comment (jopperm) [reply include/circt/Scheduling/Scheduler.h:204]: Done.

File: include/circt/Scheduling/DependenceIterator.h [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+  /// The dependence's index into the property maps, if previously assigned.
+  Optional<DependenceId> id;
+  bool operator==(const Dependence &other) const;
+};
+
+/// An iterator to transparently surface an operation's def-use dependences from
+/// the SSA subgraph (induced by the registered operations), as well as
+/// auxiliary, operation-to-operation dependences explicitly provided by the
+/// client.
+class DependenceIterator
```
[the remark this answers is not in the document]
Comment (jopperm) [reply include/circt/Scheduling/DependenceIterator.h:85]: Thanks!

File: include/circt/Scheduling/DependenceIterator.h [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+  /// The source of the dependence.
+  Operation *src;
+  /// The destination of the dependence.
+  Operation *dst;
+  /// The source operation's result number, if applicable.
+  Optional<unsigned> srcIdx;
+  /// The destination operation's operand number, if applicable.
+  Optional<unsigned> dstIdx;
+  /// The dependence's index into the property maps, if previously assigned.
+  Optional<DependenceId> id;
```
[the remark this answers is not in the document]
Comment (jopperm) [reply include/circt/Scheduling/DependenceIterator.h]: See below.

File: include/circt/Scheduling/DependenceIterator.h [hunk 13]
```diff
@@ -0,0 +1,94 @@
...
+
+namespace detail {
+
+/// Optional identifier, used to access dependence properties.
+using DependenceId = unsigned;
+
+/// A POD type to pass around the info comprising a dependence edge. This is
+/// intended to be used as a tuple with nicer element names, not for long-term
+/// storage.
+struct Dependence {
```
[the remark this answers is not in the document]
Comment (jopperm) [reply include/circt/Scheduling/DependenceIterator.h]: Done and many thanks, I love how that idea made the entire API more concise.

File: lib/Scheduling/Scheduler.cpp [hunk 14]
```diff
@@ -0,0 +1,220 @@
...
+
+  return success();
+}
+
+LogicalResult Scheduler::checkProblem() {
+  // Check the auxiliary dependences for validity, as these have been provided
+  // by the client. The def-use-dependences (from the SSA subgraph) only include
+  // registered operations by construction.
+  for (auto kv : auxDependences) {
+    for (auto &dep : getDependences(kv.first)) {
```
[the remark this answers is not in the document]
Comment (jopperm) [reply lib/Scheduling/Scheduler.cpp]: Sorry, that was an oversight. I decided to just register the endpoints of dependences provided by the client automatically, which allowed me to ditch the "is this operation registered?"-aspect of the validation altogether.

File: lib/Scheduling/Scheduler.cpp [hunk 15]
```diff
@@ -0,0 +1,220 @@
...
+  if (!getStartTime(op))
+    return op->emitError("Operation has no start time");
+  return success();
+}
+
+LogicalResult Scheduler::verifyDependence(const Dependence &dep) {
+  Operation *i = dep.src;
+  Operation *j = dep.dst;
+
+  unsigned stI, latI, stJ;
```
[the remark this answers is not in the document]
Comment (jopperm) [reply lib/Scheduling/Scheduler.cpp]: Fixed (if I understood you correctly).

**Comment by jopperm** [comment]
Thanks for the review @teqdruid! `DependenceId` is gone and replaced by a union-based `Dependence`.

> One thing to keep in mind going forward is that this latency model is simple. Like @stephenneuendorffer was saying in the discourse forum, we'd really like to see a richer latency model. (i.e. to support device primitives with non-uniform input / output cycle requirements.) I've been participating in some internal design exercises to better develop an understanding about this and how we should model state, but I have no idea (yet!) what a more complex latency model should look like.

My plan is to model offsets (for each operand/result) as a property of the operator types, for example:
```
  ┌───────────────┐
  │ ┌───┐         │
0─┼─┤   │┌────────┼─0        operator type: MAC
  │ │mul├┤ ┌───┐  │            latency:        10
1─┼─┤   │└─┤   │  │            input_offsets:  {0, 0, 6}
  │ └───┘  │add├──┼─1          output_offsets: {4, 0}
2─┼────────┤   │  │
  │        └───┘  │
  └───────────────┘
```
Here, the third operand can arrive 6 cycles after the operation is started, and the first result is available 4 cycles before the operation finishes (= start time + 10). Would that already fit your requirements?

Modeling state is a matter of imposing constraints on the binding, I think: if operations *a* and *b* share some state, then they must be bound to the same operator instance.

**Review by mikeurbach — APPROVED** [review]
Thanks again for iterating on this. I think this is a great point to land the initial implementation. It seems like you've thought far enough ahead that this isn't going to force things into an uncomfortable corner, and of course we should always be open to changes down the road anyway. I'm certainly interested in pushing on this myself in the upcoming months.

**Comment by jopperm** [comment]
Thanks @mikeurbach!

@teqdruid, if you don't have objections to the changes I made after your approval, could you merge this for me, please? (Don't have commit access yet.)

**Comment by teqdruid** [comment]
> My plan is to model offsets (for each operand/result) as a property of the operator types, for example:
> 
> ```
>   ┌───────────────┐
>   │ ┌───┐         │
> 0─┼─┤   │┌────────┼─0        operator type: MAC
>   │ │mul├┤ ┌───┐  │            latency:        10
> 1─┼─┤   │└─┤   │  │            input_offsets:  {0, 0, 6}
>   │ └───┘  │add├──┼─1          output_offsets: {4, 0}
> 2─┼────────┤   │  │
>   │        └───┘  │
>   └───────────────┘
> ```
> 
> Here, the third operand can arrive 6 cycles after the operation is started, and the first result is available 4 cycles before the operation finishes (= start time + 10). Would that already fit your requirements?

A model like that is the baseline I went into this internal design exercise with and I couldn't figure a way to model our deceptively simple computation to the level of detail I would like.

> Modeling state is a matter of imposing constraints on the binding, I think: if operations _a_ and _b_ share some state, then they must be bound to the same operator instance.

It turns out that (at least for my internal application) modeling the state and latency is both connected and more complex than that.

I'd share more, but I'll have to come up with a sufficiently different, complex example and I haven't had the chance to figure that out yet.

**Comment by teqdruid** [comment]
Very nice work! Thanks for iterating on this @jopperm.

**Comment by teqdruid** [comment]
I'll merge this once the PR gates complete.
